### PR TITLE
Partition healing

### PR DESCRIPTION
**Pull request by mennopruijssers**
This PR adds partition healing to ringpop-node. This initial PR is only healing manually. Periodic partition healing will be added in a subsequent PR.

### CODE REVIEW

**Comment by mennopruijssers** [comment]
Currently based on a "special" branch: master merged with #260 and #259

**Comment by mennopruijssers** [comment]
Manual test:
```
# create 25 node cluster
./scripts/tick-cluster.js -n 25 --interpreter node main.js

# validate no partitions:
ringpop-admin partitions $IP:3010
 08:59:18.496   Checksum    # Nodes   # Alive   # Suspect   # Faulty   Sample Host
                140365291   25        25        0           0          $IP:3010

# setup firewall using make_partition-script in ringpop-common:
sudo lsof -Pnni | ./make_partition 3000-3010 3011-3024  | sudo pfctl -emf -

# reset firewall after a while:
sudo pfctl -f /etc/pf.conf

# validate two partitions:
ringpop-admin partitions $IP:3010
 09:01:09.012   Checksum    # Nodes   # Alive   # Suspect   # Faulty   Sample Host
                228740776   14        14        0           11         $IP:3011
                883647694   11        11        0           14         $IP:3010

# run heal:
ringpop-admin heal $IP:3000
09:02:11.273 Executed heal to 1 targets
09:02:11.273  - $IP:3011
09:02:12.285 Executed heal to 1 targets
09:02:12.285  - $IP:3022
09:02:13.286 No known partitions left

# validate no partitions:
ringpop-admin partitions $IP:3010
 09:02:14.495   Checksum     # Nodes   # Alive   # Suspect   # Faulty   Sample Host
                1366565648   25        25        0           0          $IP:3010

```

**Comment by Motiejus** [comment]
Please make sure to run integration tests in travis-ci too. Appending `--enable partition-healing` to the `it-tests.js` command should do the trick here: https://github.com/uber/ringpop-node/blob/dc57cbef6a8b377daa9d24f43b25fdc20be6a5c4/test/run-shared-integration-tests#L81

File: lib/partition_healing/healer.js [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            // if the local member would become un-pingable after applying the remote change
+            if (nodeWouldBecomeUnPingable(localMember, remoteMember)) {
+                // mark it locally as suspect
+                localChanges.push(_.extend({}, remoteMember, {status: Member.Status.suspect}));
+            }
+
+            // if the remote member would become un-pingable after applying the local change
+            if (nodeWouldBecomeUnPingable(remoteMember, localMember)) {
+                // mark it remotely as suspect
+                remoteChanges.push(_.extend({}, localMember, {status: Member.Status.suspect}));
```
Comment (thanodnl) [inline lib/partition_healing/healer.js]: since the local changes are created from the `fullsync` call they contain `source` information about this node if I remember correctly. This in combination with bidirectional fullsync's caused an issue in the ringpop-node implementation.

I think we solved it by scrubbing (or never putting it on there) the source information from the change we sent remote.

If we don't do so the remote might initiate a fullsync back to this node when it receives the 'reincarnation' ping during the partition healing causing the 'bridge' between the two partitions to be created too soon.

File: lib/partition_healing/discover_provider_healer.js [hunk 2]
```diff
@@ -63,6 +63,7 @@ require('util').inherits(DiscoverProviderHealer, Healer);
  */
 DiscoverProviderHealer.prototype.heal = function heal(callback) {
     var self = this;
+    self.ringpop.stat('increment', 'heal.triggered.discover_provider');
```
Comment (thanodnl) [inline lib/partition_healing/discover_provider_healer.js]: the stat we emit here differs from the stat in `ringpop-go`: https://github.com/uber/ringpop-go/blob/dev/ringpop.go#L497 is that on purpose?

File: test/integration/partition_healing_test.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        for (var i = 0; i < partitionA.length; i++) {
+            var node = partitionA[i];
+            for (var j = 0; j < node.membership.members.length; j++) {
+                var member = node.membership.members[j];
+                if (_.pluck(partitionA, 'hostPort').indexOf(member.address) > -1) {
+                    assert.equal(member.status, 'alive')
+                } else if (_.pluck(partitionB, 'hostPort').indexOf(member.address) > -1) {
+                    assert.equal(member.status, 'faulty');
+                } else {
+                    assert.fail('!!!!');
```
Comment (thanodnl) [inline test/integration/partition_healing_test.js]: can we put a better description for the failure in the string? What is actually wrong when we reach this branch? It would make the output of the test more understandable as well as reading this test.

File: test/run-shared-integration-tests [hunk 4]
```diff
@@ -78,7 +78,7 @@ run-test-for-cluster-size() {
     # Run the tests and buffer the output to a log file. We'll display it later
     # if the test fails. This avoids interleaving of output to the terminal
     # when tests are running in parallel.
-    node "${ringpop_common_dir}/test/it-tests.js" \
+    node "${ringpop_common_dir}/test/it-tests.js" --enable-feature partition-healing \
```
Comment (thanodnl) [inline test/run-shared-integration-tests]: nit: for easier to understandable diffs now and in the future I think it is easiest if we put `--enable-feature` flags here on separate lines. It will cause only 1 addition instead of a remove and an addition when adding tests. Same when we can remove features because they became mandatory.

File: lib/partition_healing/discover_provider_healer.js (hunk 2 above)
Comment (mennopruijssers) [inline lib/partition_healing/discover_provider_healer.js]: Good point. I wanted to discuss this with @CorgiMan.
Since most partition healing code is already prepared for multiple strategies, I thought it made more sense to increment a more specific counter. @CorgiMan do you agree?

Comment (CorgiMan) [inline lib/partition_healing/discover_provider_healer.js]: I think that's fine, but let's keep the stats the same by either changing go to this style, or changing node to go's style/

Comment (mennopruijssers) [inline lib/partition_healing/discover_provider_healer.js]: See uber/ringpop-go#142

File: lib/partition_healing/healer.js [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+            // already un-pingable
+            return false;
+        }
+
+        if (Member.isStatusPingable(newState.status)) {
+            // new state is pingable
+            return false;
+        }
+
+        if (currentState.incarnationNumber > newState.incarnationNumber) {
```
Comment (CorgiMan) [inline lib/partition_healing/healer.js]: Isn't this logic done by, isNonLocalOverwrite?

Comment (CorgiMan) [inline lib/partition_healing/healer.js]: Nevermind

File: lib/partition_healing/discover_provider_healer.js [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                    }
+
+                    targets.push(target);
+                    // Remove all reachable nodes from the list of potential targets.
+                    potentialTargets = _.difference(potentialTargets, reachableNodes);
+                    next();
+                });
+            },
+            function healingDone(err) {
+                if (numberOfFailures >= self.maxNumberOfFailures) {
```
Comment (Motiejus) [inline lib/partition_healing/discover_provider_healer.js]: Do you think it would make sense to add a stat for that?

File: lib/partition_healing/healer.js [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+ */
+function Healer(ringpop) {
+    this.ringpop = ringpop;
+    this.logger = this.ringpop.loggerFactory.getLogger('healer');
+}
+
+
+/**
+ * Attempt a heal between the current and target node. Calling this function does not
+ * result in a heal ff there are nodes that need to be reincarnated to take precedence
```
Comment (Motiejus) [inline lib/partition_healing/healer.js]: s/ff/iff/ ?

File: lib/partition_healing/healer.js [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+            source: self.ringpop.whoami(),
+            incarnationNumber: self.ringpop.membership.localMember.incarnationNumber
+        }, next);
+    }
+
+    function generateChanges(joinResponse, next) {
+        remoteMembership = joinResponse.membership;
+
+        // Index local membership by address for faster lookups.
+        var localMembershipMap = _.indexBy(self.ringpop.dissemination.fullSync(), 'address');
```
Comment (Motiejus) [inline lib/partition_healing/healer.js]: In the go version, we renamed `fullSync` (equivalent to what we have here) to `MembershipAsChanges`, because it's not a full-sync anymore. Should we rename this too?

File: lib/partition_healing/healer.js [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+        var localChanges = [];
```
Comment (Motiejus) [inline lib/partition_healing/healer.js]: In the RFCs, partition healing documentation, and the public documentation, `local*` and `remote*` are called `A*` and `B*` respectively. In go, we have `changesForA` and `changesForB`.

Current way it feels a bit confusing, because just a few lines above we have `localMembershipMap`, which has a different purpose.

What do you think about renaming these two to match the Go implementation?

File: lib/partition_healing/healer.js [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+                // mark it remotely as suspect
+                remoteChanges.push(createSuspectChange(localMember));
+            }
+        }
+
+        next(null, localChanges, remoteChanges);
+    }
+
+    function createSuspectChange(member) {
+        // ignore source and sourceIncarnationNumber fields to prevent bi-directional full sync
```
Comment (Motiejus) [inline lib/partition_healing/healer.js]: Did you mean "do not send" instead of "ignore"?

Comment (mennopruijssers) [inline lib/partition_healing/healer.js]: Yeah that might be a better comment.

File: lib/partition_healing/healer.js (hunk 9 above)
Comment (mennopruijssers) [inline lib/partition_healing/healer.js]: will do.

Comment (mennopruijssers) [inline lib/partition_healing/healer.js]: I though this was actually easier to read because localChanges are applied to the local change and remoteChanges to the target (so the remote node) but it might make sense to keep them in sync.

File: lib/partition_healing/healer.js (hunk 8 above)
Comment (mennopruijssers) [inline lib/partition_healing/healer.js]: We could do it but we should do it in another pull request to keep this one as small as possible.

File: lib/partition_healing/discover_provider_healer.js (hunk 6 above)
Comment (mennopruijssers) [inline lib/partition_healing/discover_provider_healer.js]: There isn't a stat in ringpop-go. We might add it here but we have to add it to ringpop-go as well. @CorgiMan what do you think?

File: lib/partition_healing/healer.js (hunk 8 above)
Comment (mennopruijssers) [inline lib/partition_healing/healer.js]: @Motiejus #266

File: lib/partition_healing/healer.js (hunk 9 above)
Comment (mennopruijssers) [inline lib/partition_healing/healer.js]: Decided to make them in sync in the end. The only difference is that I named MA and MB membershipA and membershipB.

File: lib/partition_healing/discover_provider_healer.js (hunk 6 above)
Comment (CorgiMan) [inline lib/partition_healing/discover_provider_healer.js]: Yeah sounds like a good idea.

Comment (mennopruijssers) [inline lib/partition_healing/discover_provider_healer.js]: Cool, I'll create a separate task (and PR) for it (also for ringpop-go).

**Comment by Motiejus** [comment]
LGTM

**Comment by thanodnl** [comment]
LGTM, but make sure tests pass :)
